File: 29--Movie-Mania/Movie Mania/SearchTableViewController.m
//
//  SearchTableViewController.m
//  Movie Mania
//
//  Created by david on 11/12/15.
//  Copyright © 2015 The Iron Yard. All rights reserved.
//

#import "SearchTableViewController.h"
#import "DetailViewController.h"
#import "HistoryTableViewController.h"
#import "Movie.h"
#import "WebController.h"

@interface SearchTableViewController () <UISearchBarDelegate, UIPopoverPresentationControllerDelegate>
{
    NSMutableArray *searchResults;
    NSMutableArray *searchHistory;
    WebController *webController;
}

@property (weak, nonatomic) IBOutlet UISearchBar *searchBar;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *historyButton;

@end

@implementation SearchTableViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor colorWithHue:0.918 saturation:0.848 brightness:0.555 alpha:1];

    self.searchBar.delegate = self;
    self.historyButton.enabled = NO;
    
    UITextField *searchBarTextField = (UITextField *)[self.searchBar valueForKey:@"searchField"];
    searchBarTextField.textColor = [UIColor whiteColor];
    
    searchResults = [[NSMutableArray alloc] init];
    searchHistory = [[NSMutableArray alloc] init];
}

- (void)viewDidAppear:(BOOL)animated
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    [searchResults removeAllObjects];
    [searchHistory removeAllObjects];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [searchResults count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"SearchResultCell" forIndexPath:indexPath];
    
    Movie *movie = searchResults[indexPath.row];
    
    cell.textLabel.text = movie.title;
    cell.detailTextLabel.text = movie.year;
    cell.backgroundColor = [UIColor colorWithHue:0.918 saturation:0.848 brightness:0.555 alpha:1];;

    return cell;
}

#pragma mark - search

- (void) searchBarSearchButtonClicked:(UISearchBar *)searchBar
{
    [self search];
}

- (void) searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText
{
    [self search];
}

- (void) searchBarCancelButtonClicked:(UISearchBar *)searchBar
{
    [searchResults removeAllObjects];
    [self.tableView reloadData];
}

- (void)search
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
    
    webController = [[WebController alloc] init];
    webController.searchdelegate = self;
    [webController search:self.searchBar.text year:nil];
}

- (void)searchResultsWereFound:(NSArray *)results;
{
    [UIApplication sharedApplication].networkActivityIndicatorVisible = NO;
    
    [searchResults removeAllObjects];
    for(NSDictionary *result in results)
    {
        Movie *movie = [[Movie alloc] initSearchResultsWithDictionary:result];
        [searchResults addObject:movie];
    }
    [self.tableView reloadData];
}


#pragma mark - Navigation

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"ShowHistoryPopover"])
    {
        HistoryTableViewController *historyVC = segue.destinationViewController;
        UIPopoverPresentationController *popover = historyVC.popoverPresentationController;
        popover.delegate = self;
        historyVC.delegate = self;
        
        historyVC.history = searchHistory;
        
        historyVC.modalPresentationStyle = UIModalPresentationPopover;
        float contentSize = searchHistory.count * 44;
        
        historyVC.preferredContentSize = CGSizeMake(200, contentSize);
    }
}

- (UIModalPresentationStyle)adaptivePresentationStyleForPresentationController:(UIPresentationController *)controller
{
    return UIModalPresentationNone;
}

#pragma mark - Detail View

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    Movie *selectedMovie = searchResults[indexPath.row];
    
    self.historyButton.enabled = YES;
    [searchHistory addObject: selectedMovie];
    
    [self makeDetailVC:selectedMovie];
}

- (void)historySearchResultWasChosen:(Movie *)movie;
{
    [self makeDetailVC:movie];
}

- (void)makeDetailVC:(Movie *)selectedMovie
{
    DetailViewController *detailVC = (DetailViewController *)[self.storyboard instantiateViewControllerWithIdentifier:@"DetailViewController"];
    
    detailVC.selectedMovieTitle = selectedMovie.title;
    detailVC.selectedMovieYear = selectedMovie.year;
    
    [self.navigationController pushViewController:detailVC animated:YES];
}

@end

File: 37--Venue-Menu/Venue-Menu/Venue-Menu/SearchTableViewController.swift
//
//  SearchTableViewController.swift
//  Venue-Menu
//
//  Created by david on 11/26/15.
//  Copyright © 2015 The Iron Yard. All rights reserved.
//

import UIKit
import QuartzCore
import CoreData

protocol APIControllerProtocol
{
    func venuesWereFound(venues: [NSDictionary])
}

class SearchTableViewController: UITableViewController, APIControllerProtocol, UISearchBarDelegate
{
    @IBOutlet weak var searchBar: UISearchBar!
    @IBOutlet weak var segmentedControl: UISegmentedControl!
    
    var searchResults = [NSManagedObject]()
    var location: Location!
    
    let locationManager = LocationManager()
    var apiController: APIController!
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        segmentedControl.hidden = true
        
        searchBar.delegate = self
        location = USER_LOCATION
        
//        locationManager.delegate = self
//        locationManager.configureLocationManager()
        apiController = APIController(delegate: self)
    }
    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int
    {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return searchResults.count
    }
    
    func locationWasFound(location: Location)
    {
        self.location = location
    }
    
    @IBAction func segmentedControlValueChanged(sender: UISegmentedControl)
    {
        search()
    }
    
    func searchBarSearchButtonClicked(searchBar: UISearchBar)
    {
        search()
    }
    
    func searchBarCancelButtonClicked(searchBar: UISearchBar)
    {
        searchBar.resignFirstResponder()
    }
    
    func search()
    {
        if let _ = location
        {
            print("searching")
            print(location)
            
            if let term = searchBar.text
            {
                switch segmentedControl.selectedSegmentIndex
                {
                case 0: apiController.search(term, location: location, searchOption: "explore")
                case 1: apiController.search(term, location: location, searchOption: "search")
                default: print("segmentedControl unknown segmentIndex")
                }
            }
            
            searchBar.resignFirstResponder()
        }
    }
    
    func venuesWereFound(venues: [NSDictionary])
    {
        print(venues.count)
        dispatch_async(dispatch_get_main_queue(), {
            
            for eachVenueDict in venues
            {
                if let venue = Venue.venueWithJSON(eachVenueDict)
                {
                    self.searchResults.append(venue)
                }
            }
            self.apiController.cancelSearch()
            self.tableView.reloadData()
        })
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("SearchCell", forIndexPath: indexPath) as! VenueCell

        let venue = searchResults[indexPath.row]
        cell.venueLabel.text = venue.valueForKey("name") as? String
        cell.typeLabel.text = venue.valueForKey("type") as? String
        cell.addressLabel.text = venue.valueForKey("address") as? String
        
        if let imageURL = venue.valueForKey("icon") as? String
        {
            let imageView = makeCellImage(imageURL)
            cell.addSubview(imageView)
        }
        
        return cell
    }
    
    func makeCellImage(iconURL: String) -> UIImageView
    {
        let imageView = UIImageView()
        imageView.frame = CGRect(x: 8, y: 8, width: 64, height: 64)
        imageView.downloadedFrom(iconURL, contentMode: .ScaleToFill)
        imageView.round()
        imageView.backgroundColor = UIColor(hue:0.625, saturation:0.8, brightness:0.886, alpha:1)
        
        return imageView
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath)
    {
        let mainVC = navigationController?.viewControllers[0] as! MainViewController
        let detailVC = storyboard?.instantiateViewControllerWithIdentifier("detailVC") as! DetailViewController
        let chosenVenue = searchResults[indexPath.row]
        detailVC.venue = chosenVenue
        detailVC.location = USER_LOCATION
        detailVC.delegate = mainVC
        navigationController?.pushViewController(detailVC, animated: true)
    }
    
//    override func viewDidDisappear(animated: Bool)
//    {
//        let managedObjectContext = (UIApplication.sharedApplication().delegate as! AppDelegate).managedObjectContext
//        for venue in searchResults
//        {
//            managedObjectContext.deleteObject(venue)
//        }
//    }
}
